### PR TITLE
Add Telemetry TestCase Breakdown view

### PR DESCRIPTION
**Pull request by asankov**
Initial touches on Test Case Breakdown View.

This PR contains working Test Case Breakdown page showing some data. 

The purpose of this PR is to lay the foundations of the page. There are few questions:
1. Is this the preferred approach? - all the data is fetched via the API and the charts are drawn on the client-side, with the view returning an almost empty template that serves as a placeholder
2. Is the validation in the RPC method done right?
3. Are those the right tools to draw the charts? Patternfly wrapper of D3.js, and also some pure D3.js

The visualisation is a bit ugly now, and the colors are not right, but that can be changed easily if the other things are done right. 

Please review the draft PR and comment on what the MVP of the page is, so that I know what to focus on in the following few days.

### CODE REVIEW

**Comment by asankov** [comment]
Also, currently a proper error handling is missing when the user has no products/test plans/test cases, but I am aware of that and will fix it soon

File: tcms/settings/common.py [hunk 1]
```diff
@@ -103,9 +103,7 @@
         (_('Search Test Cases'), reverse_lazy('testcases-search')),
     ]),
     (_('TELEMETRY'), [
-        ('Coming soon',
-         'http://kiwitcms.org/blog/kiwi-tcms-team/2019/03/03/legacy-reports-become-telemetry/'),
-        ('-', '-'),
+        (_('TestCase Breakdown'), reverse_lazy('testcase-breakdown')),
```
Comment (atodorov) [inline tcms/settings/common.py]: menu is wrong, see blog

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{% extends "base.html" %}
+{% load i18n %}
+{% load static %}
+{% load comments %}
+{% load extra_filters %}
+{% load attachments_tags %}
+
+{% block title %}TestCase Breakdown{% endblock %}
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testcase-breakdown.html]: Use "Testing  Breakdown" as in the blog. Also needs to be marked for translation.

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+{% block contents %}
+
+    <div class="col-md-3">
+        Product:
+    </div>
+    <div class="col-md-3">
+        <select name="product" id="id_select_product" class="form-control selectpicker">
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testcase-breakdown.html]: Also need `<option value="">----------</option>` for Product and Test plan and the feature to query all.

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+{% load comments %}
+{% load extra_filters %}
+{% load attachments_tags %}
+
+{% block title %}TestCase Breakdown{% endblock %}
+
+{% block contents %}
+
+    <div class="col-md-3">
+        Product:
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testcase-breakdown.html]: Mark all strings for translation

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+{% load i18n %}
+{% load static %}
+{% load comments %}
+{% load extra_filters %}
+{% load attachments_tags %}
+
+{% block title %}TestCase Breakdown{% endblock %}
+
+{% block contents %}
+
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testcase-breakdown.html]: You are missing proper HTML here and this is breaking the grid and alignment. See search templates for examples.

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            {% for product in products %}
+                <option value="{{ product.pk }}">
+                    {{ product.name }}
+                </option>
+            {% endfor %}
+        </select>
+    </div>
+
+    <div class="col-md-3">
+        Test Plan:
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testcase-breakdown.html]: `Test Plan` -> `Test plan`. `:` outside of translations tag.

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+    <div class="col-md-4" style="padding: 15px; margin: 15px; text-align: center">
+        <div style="text-align: center; font-weight: bold"> Categories </div>
+        <div id="categories-chart"></div>
+    </div>
+
+    <script src="{% static 'bootstrap-select/dist/js/bootstrap-select.min.js' %}"></script>
+    <script src="{% static 'bootstrap-switch/dist/js/bootstrap-switch.min.js' %}"></script>
+    <script src="{% static "grappelli/jquery/jquery.min.js" %}"></script>
+    <script src="{% static "grappelli/js/grappelli.min.js" %}"></script>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testcase-breakdown.html]: why do you need grapelli here?

File: tcms/telemetry/urls/telemetry_urls.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+from django.conf.urls import url
+
+from tcms.telemetry.views import TestCaseBreakdownView
+
+urlpatterns = [
+    url(r'^testcase-breakdown', TestCaseBreakdownView.as_view(), name='testcase-breakdown')
+]
```
Comment (atodorov) [inline tcms/telemetry/urls/telemetry_urls.py]: Why not just urls.py ?

File: tcms/telemetry/views.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+from tcms.management.models import Product
+from tcms.testcases.models import TestCase
+from tcms.testplans.models import TestPlan
+
+
+class TestCaseBreakdownView(TemplateView):
+
+    def get(self, request, *args, **kwargs):
+        return render(request, 'telemetry/testcase-breakdown.html')
```
Comment (atodorov) [inline tcms/telemetry/views.py]: Class and template names should match the menu, not the feature name.

File: tcms/xmlrpc/api/telemetry.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
+from modernrpc.core import rpc_method
```
Comment (atodorov) [inline tcms/xmlrpc/api/telemetry.py]: Move this file under the telemetry app.

File: tcms/xmlrpc/api/telemetry.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+from tcms.management.models import Product
+from tcms.testcases.models import TestCase
+from tcms.testplans.models import TestPlan
+
+
+@rpc_method(name='Telemetry.get_data')
```
Comment (atodorov) [inline tcms/xmlrpc/api/telemetry.py]: `Testing.breakdown` to match the report name.

File: tcms/xmlrpc/api/telemetry.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+def get_data(request):
```
Comment (atodorov) [inline tcms/xmlrpc/api/telemetry.py]: `request` -> `query=None.

Also add documentation string

File: tcms/xmlrpc/api/telemetry.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+    test_plan_id = request.get('test_plan_id')
+    if not test_plan_id:
+        raise Exception('Required test_plan_id')
+
+    test_plan = TestPlan.objects.get(plan_id=test_plan_id)
+
+    test_cases = TestCase.objects.filter(plan=test_plan)
```
Comment (atodorov) [inline tcms/xmlrpc/api/telemetry.py]: Filter test cases based on `query` above. If user needs to filter by plan they can supply the appropriate field lookups. This will make the input of this method the same as `TestCase.filter` and then we can work to reuse the filtering template and JavaScript from test case search page.

File: tcms/xmlrpc/api/telemetry.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    count = {
+        'manual': 0,
+        'automated': 0
+    }
+    for test_case in test_cases:
+        if test_case.is_automated:
+            count['manual'] += 1
+        else:
+            count['automated'] += 1
+    count['all'] = count['automated'] + count['manual']
```
Comment (atodorov) [inline tcms/xmlrpc/api/telemetry.py]: Try using aggregation instead of counting by hand. The database can do this much faster:
https://docs.djangoproject.com/en/2.2/topics/db/aggregation/#generating-aggregates-for-each-item-in-a-queryset

File: tcms/xmlrpc/api/telemetry.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    categories_arr = []
+    for key, value in categories.items():
+        categories_arr.append([key, value])
+
+    statuses = {}
+    for test_case in test_cases:
+        if statuses.get(test_case.case_status.name):
+            statuses[test_case.case_status.name] += 1
+        else:
+            statuses[test_case.case_status.name] = 1
```
Comment (atodorov) [inline tcms/xmlrpc/api/telemetry.py]: Same thing for aggregation/counting for all of these fields. You can then take the values from the QuerySet and return them in the dict below.

File: tcms/telemetry/static/telemetry/js/testcase-breakdown.js [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+function updateTestPlanSelect() {
+    $('#id_select_test_plan').find('option').remove();
+
+    const productId = $('#id_select_product').val();
+    jsonRPC('TestPlan.filter', {product_id: productId}, data => {
+        updateSelect(data, '#id_select_test_plan', 'plan_id', 'name');
+        reloadCharts();
+    })
+}
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testcase-breakdown.js]: Similar methods should exist already, why duplicate them ?

File: tcms/xmlrpc/api/telemetry.py (hunk 14 above)
Comment (atodorov) [reply tcms/xmlrpc/api/telemetry.py]: Also you have a bug here. For the sample data that I use for testing I have a total of 8 test cases for the same product, which are all automated. The chart shows me all of them are manual.

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html (hunk 7 above)
Comment (asankov) [reply tcms/telemetry/templates/telemetry/testcase-breakdown.html]: For whatever reason, when I remove it I get `TypeError: $(...).selectpicker is not a function`

File: tcms/xmlrpc/api/telemetry.py (hunk 14 above)
Comment (asankov) [reply tcms/xmlrpc/api/telemetry.py]: Actually, I don't think that aggregations will work in our case. The example you are referring is counting many to many relationships for each object in the query set, while what we have is just a foreign key to another table.

For the count of automated/manual we can do it on DB level easily:
```python
test_cases = TestCase.objects.filter(**query)

count = {
    'manual': test_cases.filter(is_automated=False).count(),
    'automated': test_cases.filter(is_automated=True).count(),
    'all': test_cases.count()
}
```
for category and priority count we can do it either by hand, or with something like:
```python
test_cases = TestCase.objects.filter(**query)

priorities = {}
for priority in Priority.objects.filter():
    priorities[priority.value] = test_cases.filter(priority=priority).count()
```

File: tcms/telemetry/urls/telemetry_urls.py (hunk 8 above)
Comment (asankov) [reply tcms/telemetry/urls/telemetry_urls.py]: Because this is how it is done in other packages:
```
+ --testcase
|    +-- urls
|     |     +-- __init__.py # just imports of the other two files
|     |     +-- case_urls.py
|     |     +-- cases_urls.py
+ --testplans
|    +-- urls
|     |     +-- __init__.py # just imports of the other two files
|     |     +-- plan_urls.py
|     |     +-- plans_urls.py
```

File: tcms/xmlrpc/api/telemetry.py (hunk 10 above)
Comment (asankov) [reply tcms/xmlrpc/api/telemetry.py]: Done. What would be the appropriate name then? `tcms.telemetry.api.telemetry` looks weird.

Comment (atodorov) [reply tcms/xmlrpc/api/telemetry.py]: Just `telemetry/api.py`

File: tcms/settings/common.py [hunk 17]
```diff
@@ -103,9 +103,9 @@
         (_('Search Test Cases'), reverse_lazy('testcases-search')),
     ]),
     (_('TELEMETRY'), [
-        ('Coming soon',
```
Comment (atodorov) [inline tcms/settings/common.py]: Keep this with name 'More coming soon' at the bottom of the menu.

File: tcms/telemetry/templates/telemetry/testing-breakdown.html [hunk 18]
```diff
@@ -0,0 +1,78 @@
+{% extends "base.html" %}
+{% load i18n %}
+{% load static %}
+{% load comments %}
+{% load extra_filters %}
+{% load attachments_tags %}
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing-breakdown.html]: Don't load any template tags you are not using here.

Also rename the file as `telementry/testing/breakdown.html` to be consistent with the rest.

File: tcms/telemetry/urls/telemetry_urls.py (hunk 8 above)
Comment (atodorov) [reply tcms/telemetry/urls/telemetry_urls.py]: But `testruns/urls.py` - this is the Django way, the others are legacy.

File: tcms/telemetry/views.py [hunk 19]
```diff
@@ -0,0 +1,8 @@
+from django.shortcuts import render
+from django.views.generic import TemplateView
+
+
+class TestingBreakdownView(TemplateView):
+
+    def get(self, request, *args, **kwargs):
+        return render(request, 'telemetry/testing-breakdown.html')
```
Comment (atodorov) [inline tcms/telemetry/views.py]: I think you can configure an attribute `template_name` and nothing else.

FTR you can even pass this attribute as parameter to TemplateView in urls.py but it is ugly.

File: tcms/telemetry/urls/telemetry_urls.py [hunk 20]
```diff
@@ -0,0 +1,7 @@
+from django.conf.urls import url
+
+from tcms.telemetry.views import TestingBreakdownView
+
+urlpatterns = [
+    url(r'^testing-breakdown', TestingBreakdownView.as_view(), name='testing-breakdown')
```
Comment (atodorov) [inline tcms/telemetry/urls/telemetry_urls.py]: `testing/breakdown/$`

File: tcms/telemetry/templates/telemetry/testing-breakdown.html [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+
+{% block title %}{% trans "Testing Breakdown" %}{% endblock %}
+
+{% block contents %}
+
+    <div class="container-fluid container-cards-pf">
+
+        <h1 class="no-test-cases-heading" style="visibility: hidden;">{% trans "No test cases for this product" %}</h1>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing-breakdown.html]: What is this ?

File: tcms/telemetry/templates/telemetry/testcase-breakdown.html (hunk 7 above)
Comment (atodorov) [reply tcms/telemetry/templates/telemetry/testcase-breakdown.html]: You definitely don't need grappelli here. That problem has something to do with bootstrap-select not being loaded on time maybe.

The base template loads jquery in <head> so it should be defined. Grappelli has its own version of jQuery which you don't want to use.

File: tcms/telemetry/api/telemetry.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+    if query is None:
+        query = {}
+
+    test_cases = TestCase.objects.filter(**query)
+
+    count = {
+        'manual': test_cases.filter(is_automated=False).count(),
+        'automated': test_cases.filter(is_automated=True).count(),
+        'all': test_cases.count()
```
Comment (atodorov) [inline tcms/telemetry/api/telemetry.py]: Save the values in 2 variables and append them here. This will save a 3rd DB query.

File: tcms/telemetry/api/telemetry.py [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    priorities_qs = Priority.objects.filter().annotate(
+        num_cases=Count('priority_case')
+    ).filter(
+        num_cases__gt=0,
+        priority_case__in=test_cases
+    )
+    priorities = [[priority.value, priority.num_cases] for priority in priorities_qs]
```
Comment (atodorov) [inline tcms/telemetry/api/telemetry.py]: This is totally wrong. For 67 Test Cases which I have in the DB that results in a P1 count of over 4000.
See the query below:
`TestCase.objects.all().values('priority').annotate(num_priority=Count('priority'))`

replace `all()` with `filter(**query)`.

Everything must be coming from TestCase wherever possible.

File: tcms/telemetry/api/telemetry.py [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+
+    categories_qs = Category.objects.filter().annotate(
+        num_cases=Count('category_case')
+    ).filter(
+        num_cases__gt=0,
+        category_case__in=test_cases
+    )
+    categories = [[category.name, category.num_cases] for category in categories_qs]
```
Comment (atodorov) [inline tcms/telemetry/api/telemetry.py]:
```
TestCase.objects.all().values('category').annotate(num_priority=Count('category'))
<QuerySet [{'category': 1, 'num_priority': 46}, {'category': 2, 'num_priority': 2}, {'category': 3, 'num_priority': 2}, {'category': 4, 'num_priority': 9}, {'category': 5, 'num_priority': 8}]>
```

Mind the fact that the result is a query set (aka iterator) not a dictionary. You will have to iterate over it and convert to the expected data structure for the rest of the code to continue working.

File: tcms/telemetry/api/telemetry.py [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+
+    statuses_qs = TestCaseStatus.objects.filter().annotate(
+        num_cases=Count('testcase')
+    ).filter(
+        num_cases__gt=0,
+        testcase__in=test_cases
+    )
+    statuses = {status.name: status.num_cases for status in statuses_qs}
```
Comment (atodorov) [inline tcms/telemetry/api/telemetry.py]: Is `statuses` used ATM ?

File: tcms/telemetry/templates/telemetry/testing-breakdown.html [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+
+        <div class="col-md-3">
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing-breakdown.html]: You are still missing HTML. These divs are inside of form-group, which is inside of form.

All of the texts below should be in label tags. If you can't get it right copy testcase/search.html and remove most of it and see how it looks. The current placement of widgets is nowhere good enough compared to the search templates.

They must have the same look and feel.

File: tcms/telemetry/static/telemetry/js/testing-breakdown.js [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+});
+
+function loadInitialProduct() {
+    jsonRPC('Product.filter', {}, data => {
+        updateSelect(data, '#id_select_product', 'id', 'name');
+        updateTestPlanSelect();
+    });
+}
+
+function updateTestPlanSelect() {
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing-breakdown.js]: I am changing only Product, not touching anything else. Charts are not reloaded. 

IMO you should control how you filter the test plans based on selected product and at the end of this onchange handler simply render the charts again. This looks like to be the least complex way of doing things.

File: tcms/telemetry/static/telemetry/js/testing-breakdown.js [hunk 28]
```diff
@@ -0,0 +1,105 @@
+$(document).ready(() => {
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing-breakdown.js]: filename -> `testing/breakdown.js`

File: tcms/telemetry/api/telemetry.py (hunk 25 above)
Comment (asankov) [reply tcms/telemetry/api/telemetry.py]: Not for now. Removing them

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 29]
```diff
@@ -0,0 +1,75 @@
...
+
+        <div class="col-md-3" style="padding: 15px; margin: 15px; text-align: center">
+            <div style="text-align: center; font-weight: bold"> {% trans "Automated/ Manual" %}</div>
+            <div id="total"></div>
+            <div> {% trans "Total" %}</div>
+
+            <div id="automated-bar" style="padding: 10px; font-weight: bold"></div>
+        </div>
+
+        <div class="col-md-4" style="padding: 15px; margin: 15px; text-align: center">
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Remove styling, especially padding and margin. This should be dealth with using HTML and CSS styles on the widgets.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+            <div style="text-align: center; font-weight: bold">{% trans "Priorities" %}</div>
+
+            <div id="priorities-chart"></div>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Both Priority and Category charts must include the Y axis with numbers.

When there are the same Priorities/Categories on the X axis (extreme case only 1) and I change the product the charts are updated but the scale of them is not. So I see 50 P1 cases for one product and 3 P1 cases for another product in the same scale.  Visually this makes it very hard to detect the change.

When Y axes are shown the numbers will change and cause a flicker on the screen.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+            <svg></svg>
+        </div>
+
+        <div class="col-md-4" style="padding: 15px; margin: 15px; text-align: center">
+            <div style="text-align: center; font-weight: bold">{% trans "Categories" %}</div>
+            <div id="categories-chart"></div>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Bars should be stacked bars showing CONFIRMED/unconfirmed test cases.

There should also be a line chart on top of them showing the number of manual test cases. 

Try implementing them one by one after the rest is done. Depending on how it goes we may do this in a separate pull request.

File: tcms/telemetry/api.py [hunk 32]
```diff
@@ -0,0 +1,45 @@
...
+
+    manual_count = test_cases.filter(is_automated=False).count()
+    automated_count = test_cases.filter(is_automated=True).count()
+    count = {
+        'manual': manual_count,
+        'automated': automated_count,
+        'all': manual_count + automated_count
+    }
+
+    priorities = _get_field_count_map(test_cases, 'priority', 'priority__value')
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/telemetry/api.py]: I am not sure which would be faster either. I guess this has to be tested and verified. Should I leave it as it is until then?

[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/telemetry/api.py]: yes

File: tcms/telemetry/api.py [hunk 33]
```diff
@@ -0,0 +1,45 @@
...
+    categories = _get_field_count_map(test_cases, 'category', 'category__name')
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/telemetry/api.py]: Yes, it will look something like this:
```python
test_cases.values('priority__value', 'category__name').annotate(priority_count=Count('priority'), category_count=Count('category'))
```
however, there is a bug in this.
my current query returns the following results:
```
[['P1', 1], ['P2', 3], ['P3', 1]]
[['--default--', 4], ['other', 1]]
```
while this one returns:
```
[['P1', 1], ['P2', 2], ['P2', 1], ['P3', 1]]
[['--default--', 1], ['--default--', 2], ['other', 1], ['--default--', 1]]
```
that is because of the multiple arguments passed to `.values()`

There are 3 things I can do:
1. leave it as it is and decide whether we want to refactor it later (we are still not sure which one is the fastest)
2. come up with a smarter way to query the data
3. use the query I mentioned and aggregate the results with python code (least preferable IMO)

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 34]
```diff
@@ -0,0 +1,75 @@
...
+                        {% endfor %}
+                    </select>
+                </div>
+            </div>
+        </form>
+
+        <div class="col-md-3" style="padding: 15px; margin: 15px; text-align: center">
+            <div style="text-align: center; font-weight: bold"> {% trans "Automated/ Manual" %}</div>
+            <div id="total"></div>
+            <div> {% trans "Total" %}</div>
```
[the remark this answers is not in the document]
Comment (asankov) [reply tcms/telemetry/templates/telemetry/testing/breakdown.html]: make this a what?

[the remark this answers is not in the document]
Comment (atodorov) [reply tcms/telemetry/templates/telemetry/testing/breakdown.html]: label

File: tcms/telemetry/templates/telemetry/testing/breakdown.html (hunk 31 above)
Comment (asankov) [reply tcms/telemetry/templates/telemetry/testing/breakdown.html]: This is a lot of work, and would involve changes everywhere (API, charts). The PR is big enough now. I suggest we merge this as it is now, and this will be the next thing I will start working on

File: tcms/telemetry/static/telemetry/js/testing/breakdown.js [hunk 35]
```diff
@@ -0,0 +1,127 @@
...
+
+function updateTestPlanSelect() {
+    const productId = $('#id_select_product').val();
+
+    if (!productId) {
+        reloadCharts();
+        return;
+    }
+    jsonRPC('TestPlan.filter', {'product_id': productId}, data => {
+        updateSelect([], '#id_select_test_plan', 'plan_id', 'name');
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/breakdown.js]: `data` not empty list

File: tcms/telemetry/static/telemetry/js/testing/breakdown.js [hunk 36]
```diff
@@ -0,0 +1,127 @@
...
+        reloadCharts();
+    })
+}
+
+function populateTestPlanSelect(data = []) {
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/breakdown.js]: Remove, not used

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 37]
```diff
@@ -0,0 +1,92 @@
...
+            </div>
+
+            <div class="bar-legend">
+                <div class="bar-legend" style="display: flex; justify-content: space-around">
+                    <div style="align-items: center; display: flex;">
+                        <div style="background-color: #38A5DC; width: 20px; height: 20px;"></div>
+                        <div style="padding-left: 5px">{% trans "Automated" %}</div>
+                    </div>
+                    <div style="align-items: center; display: flex;">
+                        <div class="progress-bar-warning" style="display: block; width: 20px; height: 20px;"></div>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Remove `display: block`, see above.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 38]
```diff
@@ -0,0 +1,92 @@
...
+            <div class="progress automated-manual-bar">
+                <div class="progress-bar automated-bar" role="progressbar" data-toggle="tooltip"></div>
+                <div class="progress-bar progress-bar-warning manual-bar" role="progressbar"
+                     data-toggle="tooltip"></div>
+            </div>
+
+            <div class="bar-legend">
+                <div class="bar-legend" style="display: flex; justify-content: space-around">
+                    <div style="align-items: center; display: flex;">
+                        <div style="background-color: #38A5DC; width: 20px; height: 20px;"></div>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Don't use background-color here, use class.

File: tcms/telemetry/static/telemetry/js/testing/breakdown.js [hunk 39]
```diff
@@ -0,0 +1,122 @@
...
+        updateSelect(data, '#id_select_product', 'id', 'name');
+        reloadCharts();
+    });
+}
+
+function updateTestPlanSelect() {
+    const productId = $('#id_select_product').val();
+
+    if (!productId) {
+        reloadCharts();
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/breakdown.js]: Clear the list of selected test plans before reloading charts. To reproduce:

1) Select 1 product and see what TPs are available
2) Select ---

Result: Test plan select is not cleared.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 40]
```diff
@@ -0,0 +1,92 @@
...
+                    {% trans "Product" %}:
+                </label>
+                <div class="col-md-3">
+                    <select name="product" id="id_select_product" class="form-control selectpicker">
+                        <option value="">----------</option>
+                        {% for product in products %}
+                            <option value="{{ product.pk }}">
+                                {{ product.name }}
+                            </option>
+                        {% endfor %}
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: The backend view is not passing any initial data so this block is useless.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 41]
```diff
@@ -0,0 +1,92 @@
...
+                    {% trans "Test Plan" %}:
+                </label>
+                <div class="col-md-3">
+                    <select name="test_plan" id="id_select_test_plan" class="form-control selectpicker">
+                        <option value="">----------</option>
+                        {% for test_plan in test_plans %}
+                            <option value="{{ test_plan.pk }}">
+                                {{ test_plan.name }}
+                            </option>
+                        {% endfor %}
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: The backend view is not passing any initial data so this block is useless.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 42]
```diff
@@ -0,0 +1,92 @@
...
+    </div>
+
+    <script src="{% static 'bootstrap-select/dist/js/bootstrap-select.min.js' %}"></script>
+    <script src="{% static 'bootstrap-switch/dist/js/bootstrap-switch.min.js' %}"></script>
+
+    <script src="{% static 'js/utils.js' %}"></script>
+    <script src="{% static 'js/jsonrpc.js' %}"></script>
+
+    <script src={% static "c3/c3.min.js" %}></script>
+    <script src={% static "d3/d3.min.js" %}></script>
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Can you reorder the JS dependencies above the files coming from the application. And use consistent single quote for the static file name and double quote for the `src` value.

File: tcms/telemetry/templates/telemetry/testing/breakdown.html [hunk 43]
```diff
@@ -0,0 +1,92 @@
...
+
+    <script src="{% static 'telemetry/js/testing/breakdown.js' %}"></script>
+
+{% endblock %}
```
Comment (atodorov) [inline tcms/telemetry/templates/telemetry/testing/breakdown.html]: Missing newline at the end

File: tcms/telemetry/static/telemetry/js/testing/breakdown.js [hunk 44]
```diff
@@ -0,0 +1,122 @@
...
+
+    c3.generate(chartConfig);
+}
+
+function showOnlyRoundNumbers(number) {
+    if (number % 1 === 0) {
+        return number;
+    }
+    return '';
+}
```
Comment (atodorov) [inline tcms/telemetry/static/telemetry/js/testing/breakdown.js]: Missing newline
